### ISSUE TITLE
Does merging code coverage reports really require the source files to exist?

### ISSUE THREAD

**Issue by deviantintegral** [sebastianbergmann/php-code-coverage#487]
I was attempting to merge coverage reports on a different machine than what generated them, and I was ending up with empty reports. It looks like the filter does a file_exists() check on each file, which was failing for me.

Is there some reason we have to validate the file exists, instead of just assuming the file names are valid? If so, it would be good if here or in the phpcov CLI tool the 'merge' command documented this caveat.

**Comment by sebastianbergmann** [comment]
Yes. Once #386 is implemented it should be possible to perform the merge operation without the source files.

**Comment by cooperaj** [comment]
I've been attempting to use phpcov to merge coverage reports built in CI containers into a single report that I can push to Coveralls and have hit upon this issue. Because the source files are in a different location in the particular context that this CI task runs it's unable to find the files and is giving me empty reports.

I have looked through the code and hope to build a workaround by rewriting the file paths in the CodeCoverage object but this would be a much better solution.

**Comment by cs278** [comment]
I currently workaround this using `sed -i "s,'$(pwd)/,getcwd().'/,g" output.cov`, however changes on master are going to make this impossible: be7e0481ecf9b20e21846f4b57257c95c7796478

**Comment by cooperaj** [comment]
This is what I ended up with https://github.com/cooperaj/php-coverage-path-fixer. Shouldn't be affected too much by that change.

**Comment by cs278** [comment]
> I mean, it'll break horribly, but the technique is sound and it'll be fixable.

Unfortunately you will need to resort to reflection to get access to all of the properties that need fixing.

@sebastianbergmann Now that the coverage data is stored as serialized objects this problem is harder to workaround. Is there anything that could be added to the library to allow people to rewrite a common path prefix in one go without having to mess around with reflection?

**Comment by AndrewFeeney** [comment]
If I'm understanding the problem properly, storing the paths as relative paths as suggested in this issue https://github.com/sebastianbergmann/php-code-coverage/issues/925 might mitigate this issue for some people if all that is different is the root project path, and the files are still present on the second system. Though, I appreciate that's a less fundamental solution to the problem it might be easier to implement and solve 80% of use cases.

I'd be happy to have a look at putting in a PR if @sebastianbergmann has appetite for that?

**Comment by sebastianbergmann** [comment]
I am sorry for the messy situation that I created here.

#925 (and https://github.com/sebastianbergmann/phpcov/issues/115) are in the way of merging code coverage reports that were generated for the same code but on different systems / in different paths. The best way to address this issue, I think, would be to change the `merge()` method: find the base path for paths in `$this`, find the base path for paths in `$that`, map between the two.

To address https://github.com/sebastianbergmann/phpcov/issues/109, we need a PR that implements #874 for serializing a `ProcessedCodeCoverageData ` object to a file and vice versa without relying on `serialize()` / `unserialize()`.

To address this issue, we need a PR that implements #386, and I have honestly no idea how that problem could be tackled. Ideas, or better yet: patches, welcome.

**Comment by sebastianbergmann** [comment]
The new serializing/unserializing/merging functionality introduced in [version 14.0.0](https://github.com/sebastianbergmann/php-code-coverage/releases/tag/14.0.0) of this library does not require source files anymore.

It:

1. Unserializes coverage files
2. Validates metadata (runtime, driver, git info)
3. Calls `ProcessedCodeCoverageData::merge()` to combine line/function coverage
4. Aggregates test results

No `is_file()` checks, no source file reads. The merge operation itself is now clean.

`CodeCoverage::merge()` also still requires source files, but this is not on the common path for the use case described here. The typical workflow is: PHPUnit writes serialized coverage data, then PHPCOV merges it. This merge operation does not require source files.

While the issue no longer exists for merging, it still exists for generating reports from merged data when source files are absent.

The report generators need source files for:

- Static analysis: extracting class/trait/function info, LOC counts, executable lines, ignored lines
- Source display: HTML report shows the actual source code with coverage highlighting

To fully address this, the static analysis results as well as the source code itself would need to be serialized alongside coverage data so reports can be generated without re-reading source files. This would be a significant change that I am reluctant to pursue.

Given that the original reporter's use case (merging on a different machine) now works at the merge level, the question is whether report generation on that same machine (without source files) is also needed, or whether the merged output can be transferred back to a machine with source files for report generation.

**Comment by cs278** [comment]
Thanks 😄 

I just tested this by running the PHPUnit test suite in two different paths and then merging using a 3rd copy as the source.

The gotcha I found is the `--source` doesn't take the path I expected, I assumed I should be specifying the repository root directory but I actually had to specify into the `src` directory - presumably because that's the topmost directory in the coverage data.

```php
./phpcov merge --source /tmp/phpunit/src --html=/tmp/coverage data/
```

I found delving into the coverage data I was able to confirm this `grep --text --only-matching --perl-regexp '"basePath";s:\d+:"[^"]+"' < data/1.cov`
